packages/cubejs-api-gateway/src/gateway.ts

```diff
@@ -1207,8 +1212,8 @@ class ApiGateway {
             context,
           );
 
-          if (hasExpressionsInQuery) {
-            rewrittenQuery = this.evalMemberExpressionsInQuery(rewrittenQuery);
+          if (this.hasExpressionsInQuery(rewrittenQuery)) {
+            rewrittenQuery = this.parseMemberExpressionsInQuery(rewrittenQuery);
```

Comment:
I've not yet checked yet what can `applyRowLevelSecurity` introduce

But `parseMemberExpressionsInQuery` is not enough anyway.
There's three states of member expressions:

1. `MemberExpression`
Object with `expression` method, that takes cubes as symbols and returns SQL

2. `ParsedMemberExpression`
Object with `expression` as array of strings. Essentially JSON-compatible way of storing `MemberExpression`

3. Just a string with JSON-serialzed `ParsedMemberExpression`.

Schema compiler expects only `MemberExpression`.
`parseMemberExpressionsInQuery` turns strings in proper places in query to `ParsedMemberExpression`.
`evalMemberExpressionsInQuery` turns `ParsedMemberExpression` to `MemberExpression` in those places.

So, depending on what `applyRowLevelSecurity` can add either only `evalMemberExpressionsInQuery`, or both of `parseMemberExpressionsInQuery` + `evalMemberExpressionsInQuery` are necessary.

Reply:
`applyRowLevelSecurity` only adds `MemberExpression` objects and only in one special case: when the entire query should be denied, we add `WHERE 1 == 0`

Reply:
Correction: we actually add string member expressions: [
](https://github.com/cube-js/cube/blob/master/packages/cubejs-server-core/src/core/CompilerApi.js#L320)

Reply:
https://github.com/cube-js/cube/blob/master/packages/cubejs-server-core/src/core/CompilerApi.js#L320

Reply:
`expression: () => '1 = 0',`
That's `MemberExpression`, with `expression` as method, returnung SQL

String member expression would look like this: `query.segments.push('{"cubeName":"foo","expr": "1=0"}');`
So, for this case no need to add any additional calls after `applyRowLevelSecurity`

Reply:
Let me see if I can just add it as a parsed version and avoid this messy logic altogether

Reply:
Sorry, I've removed all my previous comment and just added the `eval`. Looks like it should be fine now